File: src/main/java/core/model/player/PlayerAvatar.java
package core.model.player;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;

/**
 *     Class to hold player avatar information
 *     Avatar is represented by overlay of graphic elements available
 *     on HT website
 */
public class PlayerAvatar {

    // Fields - Instance Variables  ====================
    private final Integer m_playerID;
    private final List<Layer> m_layers;

    // Setters and Getters ==================
    public Integer getPlayerID() {
        return m_playerID;
    }

    // Constructor and methods ;
    public PlayerAvatar(Integer playerID, String bgImage, List<Layer> layers) {
        this.m_playerID = playerID;
        this.m_layers = layers.stream().map(this::fixURL).collect(Collectors.toList());
    }


    private Layer fixURL(Layer input_layer) {
        int x = input_layer.x();
        int y = input_layer.y();
        String url = fixURL(input_layer.urlElement());

        return new Layer(x, y, url);
    }

    private String fixURL(String input_url) {

        String prefixURL = "https://www84.hattrick.org";
        if (input_url.startsWith("http")) {
            return input_url;
        }
        return prefixURL + input_url;
    }

    public void generateAvatar(Path pathAvatar) throws IOException {

        var FirstLayer = m_layers.get(0);
        int x0, y0;

        URL url = new URL(FirstLayer.urlElement());
        BufferedImage img = ImageIO.read(url.openStream());

        if (img != null) {
            BufferedImage avatar = new BufferedImage(img.getWidth(), img.getHeight(), BufferedImage.TYPE_INT_ARGB);
            Graphics g = avatar.getGraphics();

            g.drawImage(img, 0, 0, null);

            x0 = FirstLayer.x();
            y0 = FirstLayer.y();

            for (Layer layer : m_layers.stream().skip(1).toList()) {
                url = new URL(layer.urlElement());
                img = ImageIO.read(url.openStream());
                g.drawImage(img, layer.x() - x0, layer.y() - y0, null);
            }

            // Save as new image
            String pathName = pathAvatar.resolve(this.m_playerID + ".png").toString();
            ImageIO.write(avatar, "PNG", new File(pathName));
        }
    }
}

File: src/main/java/core/model/player/PlayerAvatar.kt
package core.model.player

import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.net.URL
import java.nio.file.Path
import javax.imageio.ImageIO

/**
 * Class to hold player avatar information
 * Avatar is represented by overlay of graphic elements available
 * on HT website
 */
class PlayerAvatar(val playerID: Int, imageLayers: List<Layer>) {
    private val layers: List<Layer>

    init {
        layers = imageLayers.map { inputLayer: Layer -> this.fixURL(inputLayer) }
    }

    private fun fixURL(inputLayer: Layer): Layer {
        val x = inputLayer.x
        val y = inputLayer.y
        val url = fixURL(inputLayer.urlElement)
        return Layer(x, y, url)
    }

    private fun fixURL(inputUrl: String): String {
        val prefixURL = "https://www84.hattrick.org"
        return if (inputUrl.startsWith("http")) {
            inputUrl
        } else prefixURL + inputUrl
    }

    @Throws(IOException::class)
    fun generateAvatar(pathAvatar: Path) {
        val firstLayer = layers[0]
        val x0: Int
        val y0: Int
        var url = URL(firstLayer.urlElement)
        var img = ImageIO.read(url.openStream())
        if (img != null) {
            val avatar = BufferedImage(img.width, img.height, BufferedImage.TYPE_INT_ARGB)
            val g = avatar.graphics
            g.drawImage(img, 0, 0, null)
            x0 = firstLayer.x
            y0 = firstLayer.y

            for (layer in layers.stream().skip(1).toList()) {
                url = URL(layer.urlElement)
                img = ImageIO.read(url.openStream())
                g.drawImage(img, layer.x - x0, layer.y - y0, null)
            }

            // Save as new image
            val pathName = pathAvatar.resolve("$playerID.png").toString()
            ImageIO.write(avatar, "PNG", File(pathName))
        }
    }
}
